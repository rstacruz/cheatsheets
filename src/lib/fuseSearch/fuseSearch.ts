import Fuse from 'fuse.js'
import type { SheetPage } from '~/lib/page'

/**
 * This is what gets served in searchindex.json
 */

export type ExternalSearchData = {
  index: ReturnType<FuseIndex['toJSON']>
  rows: Array<{ title: string; slug: string }>
}

export type Row = { title: string; slug: string }

export type FuseIndex = Fuse.FuseIndex<Row>

/** A subset of `SheetPage` needed for search indexing */
type PartialSheetPage = {
  slug: SheetPage['slug']
  frontmatter: Pick<SheetPage['frontmatter'], 'title'>
}

/**
 * Get `pages` and turn them into a fuse index json.
 */

export function buildFuseIndex(pages: Record<string, PartialSheetPage>) {
  const rows = Object.values(pages).map((page): Row => {
    return { title: (page.frontmatter.title ?? '') as string, slug: page.slug }
  })
  const myIndex: FuseIndex = Fuse.createIndex(['title', 'slug'], rows)
  const indexJSON = myIndex.toJSON()

  const result = { index: indexJSON, rows }
  return result
}

export async function fetchFuse() {
  const res = await fetch('/searchindex.json')
  if (res.status > 400) throw new Error('Failed to fetch searchindex.json')
  return res.json()
}

export function parseFuse(data: ExternalSearchData) {
  const index = Fuse.parseIndex(data.index)
  const fuse = new Fuse<Row>(data.rows, {}, index)
  return fuse
}
