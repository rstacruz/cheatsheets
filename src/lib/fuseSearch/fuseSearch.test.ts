import { buildFuseIndex, parseFuse } from './fuseSearch'

describe('Simple pages scenario', () => {
  // prettier-ignore
  const pages = {
    "react": { slug: "react", frontmatter: { title: "React" } },
    "vim": { slug: "vim", frontmatter: { title: "Vim" } },
    "react-router": { slug: "react-router", frontmatter: { title: "React Router" } },
  }

  let fuse: ReturnType<typeof parseFuse>

  runTest({
    label: 'one-result search',
    query: 'vim'
  })
  runTest({
    label: 'multi-result search',
    query: 'react'
  })
  runTest({
    label: 'no-result search',
    query: 'skinamarink'
  })

  beforeEach(() => {
    const externalData = buildFuseIndex(pages)
    fuse = parseFuse(externalData)
  })

  function runTest({ label, query }: { label: string; query: string }) {
    test(label, () => {
      const results = fuse.search(query)
      expect(results).toMatchSnapshot()
    })
  }
})
