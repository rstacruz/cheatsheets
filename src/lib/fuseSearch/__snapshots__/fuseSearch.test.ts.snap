// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Simple pages scenario > multi-result search 1`] = `
[
  {
    "item": {
      "slug": "react",
      "title": "React",
    },
    "refIndex": 0,
  },
  {
    "item": {
      "slug": "react-router",
      "title": "React Router",
    },
    "refIndex": 2,
  },
]
`;

exports[`Simple pages scenario > no-result search 1`] = `[]`;

exports[`Simple pages scenario > one-result search 1`] = `
[
  {
    "item": {
      "slug": "vim",
      "title": "Vim",
    },
    "refIndex": 1,
  },
]
`;
